### PR TITLE
[4.0] Drop support for WebGL 1

### LINKED ISSUE #4689: Drop support for WebGL 1 (and iOS < 15)
As of iOS 15, all supported platforms support WebGL 2.
Our renderer is littered with special cases and entirely separate rendering paths to cater to WebGL 1 limitations.
Removing them will make the code much more maintainable and adding new features much less difficult.

### PR DESCRIPTION
**Pull request by pmconne**
Fixes #4689.

I didn't modify the webgl-compatibility APIs. We should probably should. The deployed compatibility checker would need to remain on 3.x because some products will be on 3.x, others on 4.x. We should update it to notify users who lack support for WebGL 2 (chiefly iOS < 15) that 4.x products will not be usable.

I did remove direct access to `System.instance.capabilities`, to make it easier to delete all the WebGL 1-specific logic paths. Many capabilities have known minimum/true values for WebGL 2. Examples:
- max vert/frag texture units minimum 16, all we need
- max draw buffers / color attachments minimum 4, one more than we need
- 24.8 depth texture always supported, all we need

### CODE REVIEW

**Comment by pmconne** [comment]
@markschlosseratbentley do we have evidence that any devices (i.e., iPads) support WebGL 2 but do not support rendering to float nor half-float? Capabilities.ts:
```
    if (allowFloatRender && undefined !== this.queryExtensionObject("EXT_float_blend") && this.isTextureRenderable(gl, gl.FLOAT)) {
      this._maxRenderType = RenderType.TextureFloat;
    } else if (this.isWebGL2) {
      this._maxRenderType = (this.isTextureRenderable(gl, (gl as WebGL2RenderingContext).HALF_FLOAT)) ? RenderType.TextureHalfFloat : RenderType.TextureUnsignedByte;
    } else {
      const hfExt: OES_texture_half_float | undefined = this.queryExtensionObject<OES_texture_half_float>("OES_texture_half_float");
      this._maxRenderType = (hfExt !== undefined && this.isTextureRenderable(gl, hfExt.HALF_FLOAT_OES)) ? RenderType.TextureHalfFloat : RenderType.TextureUnsignedByte;
    }
```

**Comment by pmconne** [comment]
@markschlosseratbentley 

```this._canRenderDepthWithoutColor = this._maxDepthType === DepthType.TextureUnsignedInt24Stencil8 ? this.isDepthRenderableWithoutColor(gl) : false;
```

maxDepthType is guaranteed to be 24.8 for webgl 2.

```
  private isDepthRenderableWithoutColor(gl: WebGLContext): boolean {
    const dtExt = this.queryExtensionObject<WEBGL_depth_texture>("WEBGL_depth_texture");
    if (dtExt === undefined)
      return false;
```

The `WEBGL_depth_texture` extension was integrated into WebGL 2 core. Do we have evidence of contexts that support 24.8 depth texture but not rendering to it without also binding a color attachment?

**Comment by markschlosseratbentley** [comment]
> maxDepthType is guaranteed to be 24.8 for webgl 2.
> 
> The `WEBGL_depth_texture` extension was integrated into WebGL 2 core. Do we have evidence of contexts that support 24.8 depth texture but not rendering to it without also binding a color attachment?
> 
> EDIT: `canRenderDepthWithoutColor` is referenced only in SolarShadowMap.ts, with a comment about MacOS Safari. Was that with WebGL 1, or 2? Any details?

Just to update this; I'm digging through notes to see if I can find any more details.  Currently unsure.

**Comment by pmconne** [comment]
Awaiting ImageTests results.

**Comment by markschlosseratbentley** [comment]
> > maxDepthType is guaranteed to be 24.8 for webgl 2.
> > The `WEBGL_depth_texture` extension was integrated into WebGL 2 core. Do we have evidence of contexts that support 24.8 depth texture but not rendering to it without also binding a color attachment?
> > EDIT: `canRenderDepthWithoutColor` is referenced only in SolarShadowMap.ts, with a comment about MacOS Safari. Was that with WebGL 1, or 2? Any details?
> 
> Just to update this; I'm digging through notes to see if I can find any more details. Currently unsure.

Found [original issue](https://dev.azure.com/bentleycs/iModelTechnologies/_workitems/edit/254521/) + [PR](https://dev.azure.com/bentleycs/iModelTechnologies/_git/imodeljs/pullrequest/63498).

Not sure from those if the original issue was for WebGL 1 or 2 or both.  I can try it out on modern MacOS Safari and see if it is still needed.

File: core/frontend/src/render/webgl/GLTimer.ts [hunk 1]
```diff
@@ -8,59 +8,13 @@ import { IModelError } from "@itwin/core-common";
...
-    super();
-    if (system.capabilities.isWebGL2) {
-      this._e = system.capabilities.queryExtensionObject<any>("EXT_disjoint_timer_query_webgl2");
-      if (this._e === undefined) // If webgl2 timer doesn't work, attempt to use the older disjoint timer
-        this._e = system.capabilities.queryExtensionObject<any>("EXT_disjoint_timer_query");
-    } else
-      this._e = undefined;
-    this._context = system.context as WebGL2RenderingContext;
+    this._e = system.disjointTimerQuery;
+    this._context = system.context;
```
Comment (DStradley) [inline core/frontend/src/render/webgl/GLTimer.ts:17]: A minor question, do we not need any longer to drop back to EXT_disjoint_timer_query if EXT_disjoint_timer_query_webgl2 fails?  I see that 5 years ago, FireFox only supported EXT_disjoint_timer_query under webgl2, but that likely has changed, so maybe it does not matter now?

Comment (pmconne) [reply core/frontend/src/render/webgl/GLTimer.ts:17]: I do not know if anything has changed about FF's support for that extension. Something to consider pursuing separately.

**Comment by markschlosseratbentley** [comment]
> > > maxDepthType is guaranteed to be 24.8 for webgl 2.
> > > The `WEBGL_depth_texture` extension was integrated into WebGL 2 core. Do we have evidence of contexts that support 24.8 depth texture but not rendering to it without also binding a color attachment?
> > > EDIT: `canRenderDepthWithoutColor` is referenced only in SolarShadowMap.ts, with a comment about MacOS Safari. Was that with WebGL 1, or 2? Any details?
> > 
> > 
> > Just to update this; I'm digging through notes to see if I can find any more details. Currently unsure.
> 
> Found [original issue](https://dev.azure.com/bentleycs/iModelTechnologies/_workitems/edit/254521/) + [PR](https://dev.azure.com/bentleycs/iModelTechnologies/_git/imodeljs/pullrequest/63498).
> 
> Not sure from those if the original issue was for WebGL 1 or 2 or both.  I can try it out on modern MacOS Safari and see if it is still needed.

On M1 Mac Mini, fully updated, the shadow problem no longer occurs in Safari with the `canRenderDepthWithoutColor` workaround removed - both using WebGL1 and 2.

It is possible the issue never occurred on M1 Mac MIni, however.

The original issue likely was on Intel Mac Mini.  I will try to boot that up and try it out.  It may take some time.

File: core/frontend/src/render/webgl/GLTimer.ts (hunk 1 above)
Comment (pmconne) [reply core/frontend/src/render/webgl/GLTimer.ts:17]: Thanks. Please file an issue if you think we should try to find out if it's still reproducible.

**Comment by pmconne** [comment]
>  I will try to boot that up and try it out.

Thanks. If you are able to determine we no longer need the workaround, we should remove it in a separate PR.

**Review by DStradley — APPROVED** [review]

**Pull request merged**